### PR TITLE
Add `implement_via_object` to `rustc_deny_explicit_impl` to control object candidate assembly

### PR DESCRIPTION
**Pull request by compiler-errors**
Some built-in traits are special, since they are used to prove facts about the program that are important for later phases of compilation such as codegen and CTFE. For example, the `Unsize` trait is used to assert to the compiler that we are able to unsize a type into another type. It doesn't have any methods because it doesn't actually *instruct* the compiler how to do this unsizing, but this is later used (alongside an exhaustive match of combinations of unsizeable types) during codegen to generate unsize coercion code.

Due to this, these built-in traits are incompatible with the type erasure provided by object types. For example, the existence of `dyn Unsize<T>` does not mean that the compiler is able to unsize `Box<dyn Unsize<T>>` into `Box<T>`, since `Unsize` is a *witness* to the fact that a type can be unsized, and it doesn't actually encode that unsizing operation in its vtable as mentioned above.

The old trait solver gets around this fact by having complex control flow that never considers object bounds for certain built-in traits:
https://github.com/rust-lang/rust/blob/2f896da247e0ee8f0bef7cd7c54cfbea255b9f68/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L61-L132

However, candidate assembly in the new solver is much more lovely, and I'd hate to add this list of opt-out cases into the new solver. Instead of maintaining this complex and hard-coded control flow, instead we can make this a property of the trait via a built-in attribute. We already have such a build attribute that's applied to every single trait that we care about: `rustc_deny_explicit_impl`. This PR adds `implement_via_object` as a meta-item to that attribute that allows us to opt a trait out of object-bound candidate assembly as well.

r? @lcnr

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/assembly/mod.rs [hunk 1]
```diff
@@ -631,6 +632,13 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         goal: Goal<'tcx, G>,
         candidates: &mut Vec<Candidate<'tcx>>,
     ) {
+        if self.tcx().has_attr(
+            goal.predicate.trait_def_id(self.tcx()),
+            sym::rustc_do_not_implement_via_object,
+        ) {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/assembly/mod.rs:638]: please instead add another field to `TraitDef` and access that here. you can then mark `rustc_do_not_implement_via_object` as `@only_local`.

**Review by lcnr — COMMENTED** [review]
as a thought, do we want to change this to instead add a argument to `rustc_deny_explicit_impl`, i.e. `#[rustc_deny_explicit_impl(builtin_object_impl = <bool>)]` :thinking: 

going with opt-out here feels dangerous to me as getting it wrong is unsound(?), so maybe alternatively change it to `#[rustc_builtin_has_object_impl]`? :thinking:

**Comment by compiler-errors** [comment]
> as a thought, do we want to change this to instead add a argument to rustc_deny_explicit_impl, i.e. #[rustc_deny_explicit_impl(builtin_object_impl = <bool>)] 🤔

I don't think we can do that, since the transmutability trait `BikeshedIntrinsicFrom`, I *think* will allow custom implementations in order to indicate that a type is transmutable to another trait... 🤔 Maybe I'm wrong, though, cc @jswrenn / @bryangarza.

**Review by lcnr — COMMENTED** [review]
pls add check to `rustc_passes/src/check_attr.rs`, after that r=me

**Comment by bryangarza** [comment]
> > as a thought, do we want to change this to instead add a argument to rustc_deny_explicit_impl, i.e. #[rustc_deny_explicit_impl(builtin_object_impl = )] 🤔
> 
> I don't think we can do that, since the transmutability trait `BikeshedIntrinsicFrom`, I _think_ will allow custom implementations in order to indicate that a type is transmutable to another trait... 🤔 Maybe I'm wrong, though, cc @jswrenn / @bryangarza.

`BikeshedIntrinsicFrom` is not meant to be implemented by end-users, I *think* the only way they can use it is by using it as as trait bound, and then the trait solver will check that `BikeshedIntrinsicFrom<Src> is implemented for `Dst`

If you're asking about custom implementations in rustc itself... I can check with @jswrenn offline and get back to you

File: compiler/rustc_middle/src/ty/trait_def.rs [hunk 2]
```diff
@@ -52,6 +52,14 @@ pub struct TraitDef {
     /// List of functions from `#[rustc_must_implement_one_of]` attribute one of which
     /// must be implemented.
     pub must_implement_one_of: Option<Box<[Ident]>>,
+
+    /// Whether a type's built-in `dyn Trait: Trait` implementation is explicitly
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/trait_def.rs:56]: I'm talking about the built-in fact that every `dyn Trait` implements `Trait`

**Comment by compiler-errors** [comment]
Well, given that `BikeshedIntrinsicFrom` is not intended to be implemented (afaict, could change later I guess, it *is* an unsafe trait for a reason perhaps?) I just made this an attribute on `rustc_deny_explicit_impl`.

Maybe we could uplift this into something like `rustc_builtin_impl(deny_explicit_impl = true, deny_builtin_object_impl = true)` but I'm too lazy to do that, could be convinced otherwise tho I guess :smile_cat:

**Review by lcnr — COMMENTED** [review]
r=me after nits and updating PR descr + title

**Pull request merged**